### ISSUE TITLE
Some setuptools commands are deprecated (and others will also eventually be)

### ISSUE THREAD

**Issue by abravalheri** [devpi/devpi#870]
Hello devpi maintainers,

Recently `setuptools` has started to fully embrace PEP 517 and any direct invocation of `python setup.py` is (in general) [considered deprecated](https://setuptools.pypa.io/en/latest/history.html#v58-3-0) (some commands have already been officially deprecated, while others will eventually be).

Moreover `sphinx` [has already deprecated `build_sphinx`](https://github.com/sphinx-doc/sphinx/pull/10040).

When recently discussing these deprecations on the setuptools repository, I noticed that `devpi` seems to still use quite a bit of these deprecated (or soon to be deprecated) commands ([there are mentions here for example](https://github.com/devpi/devpi/blob/7b2a40e87e9927a28eb554d0454130cd663e34f3/client/devpi/upload.py)).

Is my impression correct? (I just wanted to give a heads up so we can avoid disruption when the time comes...)

Is there a list of setuptools commands used by `devpi` somewhere? Maybe we can help to devise a plan or find alternatives to make this transition as smooth as possible...

**Comment by fschulze** [comment]
Thanks for creating a ticket which summarizes the deprecations. I was aware of them, but didn't know the timeline and hadn't researched the replacements yet. There is a major version of devpi-client upcoming and I will include the necessary changes in that.

**Comment by abravalheri** [comment]
That is awesome @fschulze , thank you for working on this!

In general for building packages the recommendation is to use `python -m build` (which should work as a drop in replacement for `python setup.py sdist/bdist_wheel`, but if it doesn't please let setuptools know so we can work on the problem).

Regarding `build_sphinx`, there is no real standardisation in the area, so it is difficult to guess what would be the best approach.
In general, I would say that if the package has a `docs/conf.py`, the client could try running `sphinx-build` directly...

Regarding `upload_docs`, it would be nice to create a strategy on how to replace the command (although it is not deprecated yet, there is some intention to remove it and making devpi-client independent of it would help setuptools to move forward).

Installations can be handed over to pip or https://github.com/pradyunsg/installer. 

Finally I suppose the regular uploads are already handled by twine... Is there any other setuptools command devpi-client is using that I forgot to mention?

**Comment by fschulze** [comment]
@abravalheri So far I replaced ``setup.py sdist/bdist_wheel``.

The upload is done directly via requests. I might switch to twine, not sure yet.

I'm currently working on the documentation building. The only thing I didn't find a replacement for yet is ``python setup.py --name`` and ``python setup.py --version``. It is currently used to create the name of the ``.doc.zip`` file. This is the name and version of the package, not the docs from ``conf.py``. I think this might need to be in ``build`` to support all the various ways a package can be setup nowadays (setup.py, setup.cfg, pyproject.toml, etc).

**Comment by abravalheri** [comment]
Thank you @fschulze!

I noticed that some libraries are using [pep517.meta](https://github.com/pypa/pep517/blob/a942316305395f8f757f210e2b16f738af73f8b8/pep517/meta.py#L63) to load the metadata.

Something I don't know is how the authors of the library are handling the split with `build`, and if one of the two libraries want to keep this piece of API public. Maybe it is something worth checking?

I did a [quick search](https://github.com/search?l=Python&q=pep517.meta&type=Code), and there is definitely a bunch of tools out there relying on it, including widespread ones like poetry and towncrier. So that might be an argument for them keeping it public?
